### PR TITLE
Introduce SegGPT model

### PR DESCRIPTION
**Pull request by raghavanone**
Introduce SegGPT model.

#27514

### CODE REVIEW

**Comment by ArthurZucker** [comment]
cc @amyeroberts and @niels

**Comment by amyeroberts** [comment]
Hi @raghavanone - thanks for opening this PR! Let us know when it's ready for review

File: src/transformers/models/seggpt/modeling_seggpt.py [hunk 1]
```diff
@@ -0,0 +1,732 @@
...
+        if not return_dict:
+            outputs = (
+                logits,
+                encoder_outputs[0],
+            )
+            outputs = outputs + (encoder_outputs[1],) if output_hidden_states else outputs
+            return outputs + (encoder_outputs[-1],) if output_attentions else outputs
+
+        return SegGPTModelOutput(
+            logits=logits,
```
Comment (NielsRogge) [inline src/transformers/models/seggpt/modeling_seggpt.py:676]: base models should not output logits. Only head models.

File: src/transformers/models/seggpt/modeling_seggpt.py [hunk 2]
```diff
@@ -0,0 +1,732 @@
...
+
+
+class SegGPTModel(SegGPTPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+
+        # --------------------------------------------------------------------------
+        self.encoder = SegGPTEncoder(config)
+
+        self.decoder = SegGPTDecoder(config)
```
Comment (NielsRogge) [inline src/transformers/models/seggpt/modeling_seggpt.py:595]: The decoder needs to be part of the head model. `SegGPTModel` should only output `last_hidden_state`, `hidden_states` and `attentions` (similar to `BertModel`, `Gpt2Model`, etc.).

File: src/transformers/models/seggpt/modeling_seggpt.py [hunk 3]
```diff
@@ -0,0 +1,732 @@
...
+            hidden_states=encoder_outputs[1] if output_hidden_states else None,
+            attentions=encoder_outputs[-1] if output_attentions else None,
+        )
+
+
+class SegGPTForInstanceSegmentation(SegGPTPreTrainedModel):
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.seggpt_model = SegGPTModel(config)
```
Comment (NielsRogge) [inline src/transformers/models/seggpt/modeling_seggpt.py:687]: decoder can be added here

**Comment by EduardoPach** [comment]
@raghavanone Hey, I also was doing the implementation of `SegGPT` to add to the library (you can find my work here https://github.com/huggingface/transformers/pull/27735) I believe my current implementation follows a bit more the library standards (@NielsRogge if you could take a look at my PR to make sure my claim is right). If you want to we can collaborate on my PR

**Comment by raghavanone** [comment]
@EduardoPach my implementations is close to completion,the test pass and conversion script is also complete.I don't think it is wise to abandon this.

File: src/transformers/models/seggpt/modeling_seggpt.py (hunk 1 above)
Comment (raghavanone) [reply src/transformers/models/seggpt/modeling_seggpt.py]: Understood, Done.

File: src/transformers/models/seggpt/modeling_seggpt.py (hunk 2 above)
Comment (raghavanone) [reply src/transformers/models/seggpt/modeling_seggpt.py]: Done.

**Comment by raghavanone** [comment]
@amyeroberts  Requesting for first round review.
